### PR TITLE
[Five 116] Crear servicio de Categorias

### PR DESCRIPTION
**Pull request by gperelstein**
Descripción
Se debe crear un servicio de categorias que realice la comunicación con la base de datos.

Cambios realizados:

- Se creó la interfaz ICategoriesService con los métodos GetAll, Get, Save, Update y Delete

- Se creó el servicio CategoriesService con los métodos GetAll, Get, Save, Update y Delete

- Se hizo la inyección de dependecias del CategoriesService en el Startup

- Se modificaron los modelos de Category y ProjectCategory para relacionar las categorías con los proyectos

### CODE REVIEW

File: FiveRockingFingers/FRF.Core/AutoMapperProfile.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ public AutoMapperProfile()
             CreateMap<DataAccess.EntityModels.Project, Models.Project>()
                 .ReverseMap().ForMember(dest => dest.Id, act => act.Ignore()).ForMember(dest => dest.ProjectCategories, act => act.Ignore());
             CreateMap<DataAccess.EntityModels.Category, Models.Category>()
-                .ReverseMap();
+                .ReverseMap().ForMember(dest => dest.Id, act => act.Ignore()).ForMember(dest => dest.ProjectCategories, act => act.Ignore());
```
Comment (matiasbeckerle) [inline FiveRockingFingers/FRF.Core/AutoMapperProfile.cs:17]: Please use one line per statement so it gets easier to read.

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace FRF.Core.Services
+{
+    public class CategoriesService : ICategoriesService
+    {
+        private readonly IConfiguration _configuration;
+        private readonly DataAccessContext _dataContext;
+        private readonly IMapper _mapper;
```
Comment (matiasbeckerle) [inline FiveRockingFingers/FRF.Core/Services/CategoriesService.cs:16]: Please use one empty line between this line and constructor of the class.

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        public CategoriesService(IConfiguration configuration, DataAccessContext dataContext, IMapper mapper)
+        {
+            _configuration = configuration;
+            _dataContext = dataContext;
+            _mapper = mapper;
+        }
+
+        public async Task<List<Category>> GetAll()
+        {
+            var result = await _dataContext.Categories.Include(c => c.ProjectCategories).ThenInclude(pc => pc.Project).ToListAsync();
```
Comment (matiasbeckerle) [inline FiveRockingFingers/FRF.Core/Services/CategoriesService.cs:26]: Consider changing the style to...

```
var result = await _dataContext.Categories
  .Include(c => c.ProjectCategories)
    .ThenInclude(pc => pc.Project)
  .ToListAsync();
```

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        private readonly DataAccessContext _dataContext;
+        private readonly IMapper _mapper;
+        public CategoriesService(IConfiguration configuration, DataAccessContext dataContext, IMapper mapper)
+        {
+            _configuration = configuration;
+            _dataContext = dataContext;
+            _mapper = mapper;
+        }
+
+        public async Task<List<Category>> GetAll()
```
Comment (matiasbeckerle) [inline FiveRockingFingers/FRF.Core/Services/CategoriesService.cs:24]: Missing async suffix.

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            var result = await _dataContext.Categories.Include(c => c.ProjectCategories).ThenInclude(pc => pc.Project).ToListAsync();
+            if (result == null)
+            {
+                return null;
+            }
```
Comment (matiasbeckerle) [inline FiveRockingFingers/FRF.Core/Services/CategoriesService.cs:30]: Does this ever returns null? Isn't the query returning an empty list instead?

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            return _mapper.Map<List<Category>>(result);
+        }
+
+        public async Task<Category> Get(int id)
```
Comment (matiasbeckerle) [inline FiveRockingFingers/FRF.Core/Services/CategoriesService.cs:34]: Same all comments as above.

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        public async Task<Category> Update(Category category)
+        {
+            var result = await _dataContext.Categories.Include(c => c.ProjectCategories).SingleAsync(c => c.Id == category.Id);
+
+            if (result == null)
+            {
+                throw new System.ArgumentException("There is no category with Id = " + category.Id, "category.Id");
+            }
```
Comment (matiasbeckerle) [inline FiveRockingFingers/FRF.Core/Services/CategoriesService.cs:65]: I think you could instead return null and answer 404 on the controller.

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            return;
+        }
+
+        
+
+        
+
+        
+
+
```
Comment (matiasbeckerle) [inline FiveRockingFingers/FRF.Core/Services/CategoriesService.cs:89]: Hehehe too many empty lines.

File: FiveRockingFingers/FRF.Core/AutoMapperProfile.cs (hunk 1 above)
Comment (gperelstein) [reply FiveRockingFingers/FRF.Core/AutoMapperProfile.cs]: Done!

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs (hunk 2 above)
Comment (gperelstein) [reply FiveRockingFingers/FRF.Core/Services/CategoriesService.cs:16]: Done!

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs (hunk 3 above)
Comment (gperelstein) [reply FiveRockingFingers/FRF.Core/Services/CategoriesService.cs]: Done!

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs (hunk 4 above)
Comment (gperelstein) [reply FiveRockingFingers/FRF.Core/Services/CategoriesService.cs]: Done!

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs (hunk 5 above)
Comment (gperelstein) [reply FiveRockingFingers/FRF.Core/Services/CategoriesService.cs]: You were right! This 'if' was unnecessary

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs (hunk 6 above)
Comment (gperelstein) [reply FiveRockingFingers/FRF.Core/Services/CategoriesService.cs]: Done!

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs (hunk 7 above)
Comment (gperelstein) [reply FiveRockingFingers/FRF.Core/Services/CategoriesService.cs]: You are right! The controller was already doing that so this 'if' was unnecessary

File: FiveRockingFingers/FRF.Core/Services/CategoriesService.cs (hunk 8 above)
Comment (gperelstein) [reply FiveRockingFingers/FRF.Core/Services/CategoriesService.cs]: Hehehe I didn't saw them. I already fix it

**Review by matiasbeckerle — APPROVED** [review]

**Pull request merged**